src/beatmap.c

```diff
@@ -185,6 +185,28 @@ static void parse_one_hit(char *line, struct oshu_hit **hit)
 }
 
 /**
+ * Parse #oshu_hit::combo and #oshu_hit::combo_seq of a single #oshu_hit
+ */
+static void parse_hit_combo(struct oshu_hit *previous, struct oshu_hit *hit)
+{
+	int new_combo = hit->type & OSHU_HIT_NEW_COMBO;
+	int skip_combo = (hit->type & OSHU_HIT_COMBO_MASK) >> 4;
+
+	if (previous != NULL) {
+		if (new_combo) {
+			hit->combo = previous->combo + 1 + skip_combo;
+			hit->combo_seq = 0;
+		} else {
+			hit->combo_seq = previous->combo_seq + 1;
+			hit->combo = previous->combo;
+		}
+	} else {
+		hit->combo = 0;
+		hit->combo_seq = 0;
```

Comment:
The first `combo_seq` is actually 1 if we follow the documentation of the field. It's kinda counter-intuitive for a developer, but this is what the user sees.